src/System.Device.Gpio/System/Device/Gpio/Drivers/LibGpiodDriver.Linux.cs

```diff
@@ -0,0 +1,279 @@
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Runtime.InteropServices;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace System.Device.Gpio.Drivers
+{
+    public class LibGpiodDriver : UnixDriver
+    {
+        private string _deviceName = "gpiochip0";
+
+        private SafeChipHandle _chip;
+
+        private Dictionary<int, SafeLineHandle> _pinNumberToSafeLineHandle;
+
+        private timespec _defaultTimeOut = new timespec(IntPtr.Zero, new IntPtr(1_000_000)); // one millisecond
+
+        private Dictionary<int, LibGpiodDriverEventHandler> _pinNumberToEventHandler = new Dictionary<int, LibGpiodDriverEventHandler>();
+
+        protected internal override int PinCount => Interop.gpiod_chip_num_lines(_chip);
+
+        public LibGpiodDriver()
+        {
+            OpenChip();
+            _pinNumberToSafeLineHandle = new Dictionary<int, SafeLineHandle>(PinCount);
+        }
+
+        private void OpenChip()
+        {
+            SafeChipIteratorHandle iterator = Interop.gpiod_chip_iter_new();
+            if (iterator == null)
+            {
+                throw new IOException($"No any chip available, error code: {Marshal.GetLastWin32Error()}");
+            }
+            _chip = Interop.gpiod_chip_iter_next(iterator);
+            if (_chip == null)
+            {
+                throw new IOException($"No chip found, error code: {Marshal.GetLastWin32Error()}");
+            }
+            _deviceName = Marshal.PtrToStringAnsi(Interop.gpiod_chip_name(_chip));
+            // Freeing other chips opened
+            Interop.gpiod_chip_iter_free_noclose(iterator);
+        }
+
+        private void RequestEvent(int pinNumber, LibGpiodDriverEventHandler eventDescriptor, PinEventTypes eventType)
+        {
+            int eventSuccess = -1;
+            SafeLineHandle pin = Interop.gpiod_chip_get_line(_chip, pinNumber);
+            if (pin != null)
+            {
+                eventDescriptor.PinHandle = pin;
+                if (eventType == PinEventTypes.Rising || eventType == PinEventTypes.Falling)
+                {
+                    eventSuccess = Interop.gpiod_line_request_both_edges_events(pin, $"Listen {pinNumber} for falling edge event");
+                }
+                else
+                {
+                    throw new ArgumentException("Invalid or not supported event type requested");
+                }
+            }
+            if (eventSuccess < 0)
+            {
+                throw new IOException($"Error while requesting event listener for pin {pinNumber}, error code: {Marshal.GetLastWin32Error()}");
+            }
+        }
+
+        protected internal override void AddCallbackForPinValueChangedEvent(int pinNumber, PinEventTypes eventType, PinChangeEventHandler callback)
+        {
+            if (!_pinNumberToEventHandler.TryGetValue(pinNumber, out LibGpiodDriverEventHandler eventHandler))
```

Comment:
If you are locking access to this you should lock access everywhere. Such as here. It is not safe to read and write a dictionaryconcurrently

Reply:
Initially i put lock was like this:
```csharp
lock (_pinNumberToEventHandler)
 {
            if (!_pinNumberToEventHandler.TryGetValue(pinNumber, out LibGpiodDriverEventHandler eventHandler))
            {
                if (_pinNumberToSafeLineHandle.TryGetValue(pinNumber, out SafeLineHandle pin))
                    Interop.ReleaseGpiodLine(pin);
                eventHandler = new LibGpiodDriverEventHandler(pinNumber, new CancellationTokenSource());
                _pinNumberToEventHandler.Add(pinNumber, eventHandler);
                }
                eventHandler.PinHandle = RequestEvent(pinNumber, eventType);
                InitializeEventDetectionTask(eventHandler);
            } 
}
```
but felt its too broad locking, was not able to figure out better granular approach and ended up with this. Any suggestions for that or are you ok with this old approach with broader lock?

Reply:
if we going to lock with every read/write operation from the dictionary then we should use ConcurrentDictionary

Reply:
@krwq are you OK with using concurrent dictionary